src/sql/src/query_model/mod.rs

```diff
@@ -0,0 +1,407 @@
+        self.make_box(BoxType::Select(Select::new()))
+    }
+
+    fn get_box(&self, box_id: BoxId) -> &RefCell<QueryBox> {
+        &*self.boxes.get(&box_id).unwrap()
+    }
+
+    /// Create a new quantifier and adds it to the parent box
+    fn make_quantifier(
+        &mut self,
+        quantifier_type: QuantifierType,
+        input_box: BoxId,
+        parent_box: BoxId,
+    ) -> QuantifierId {
+        let id = self.next_quantifier_id;
+        self.next_quantifier_id += 1;
+        let q = Box::new(RefCell::new(Quantifier {
+            id,
+            quantifier_type,
+            input_box,
+            parent_box,
+            alias: None,
+        }));
+        self.quantifiers.insert(id, q);
+        self.get_box(parent_box).borrow_mut().quantifiers.insert(id);
+        self.get_box(input_box)
+            .borrow_mut()
+            .ranging_quantifiers
+            .insert(id);
+        id
+    }
+
+    fn get_quantifier(&self, quantifier_id: BoxId) -> &RefCell<Quantifier> {
+        &*self.quantifiers.get(&quantifier_id).unwrap()
+    }
+
+    /// Visit boxes in the query graph in pre-order
+    fn visit_pre_boxes<F, E>(&self, f: &mut F) -> Result<(), E>
+    where
+        F: FnMut(&RefCell<QueryBox>) -> Result<(), E>,
+    {
+        self.visit_pre_boxes_in_subgraph(f, self.top_box)
+    }
+
+    /// Visit boxes in the query graph in pre-order
+    fn visit_pre_boxes_in_subgraph<F, E>(&self, f: &mut F, start_box: BoxId) -> Result<(), E>
+    where
+        F: FnMut(&RefCell<QueryBox>) -> Result<(), E>,
+    {
+        let mut visited = HashSet::new();
+        let mut stack = vec![start_box];
+        while !stack.is_empty() {
+            let box_id = stack.pop().unwrap();
+            if visited.insert(box_id) {
+                let query_box = self.get_box(box_id);
+                f(query_box)?;
+
+                stack.extend(
+                    query_box
+                        .borrow()
+                        .quantifiers
+                        .iter()
+                        .rev()
+                        .map(|q| self.get_quantifier(*q).borrow().input_box),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Removes unreferenced objects from the model. May be invoked
+    /// several times during query rewrites.
+    fn gargabe_collect(&mut self) {
+        let mut visited_boxes = HashSet::new();
+        let mut visited_quantifiers = HashSet::new();
+        let mut stack = vec![self.top_box];
+        while !stack.is_empty() {
+            let box_id = stack.pop().unwrap();
+            if visited_boxes.insert(box_id) {
+                let query_box = self.get_box(box_id);
+                for q in query_box.borrow().quantifiers.iter() {
+                    visited_quantifiers.insert(*q);
+                    let q = self.get_quantifier(*q).borrow();
+                    stack.push(q.input_box);
+                }
+            }
+        }
```

Comment:
Sure. I duplicated the code since `gabage_collect` needs access to the set of visited boxes. I could pass it as a mutable reference to `visit_pre_boxes_in_subgraph`, perhaps adding a new override for it.